### PR TITLE
Upgrade go.mod files to Go 1.20

### PR DESCRIPTION
**Pull request by clfs**
#### Summary

Addresses https://github.com/sigstore/sigstore/issues/1063:

>Since PR https://github.com/sigstore/sigstore/pull/962, CI runs as Go 1.20: 
>
>https://github.com/sigstore/sigstore/blob/589100556a55096e80e6b3c0ca6028da6fb9adbb/.github/workflows/e2e_test.yml#L25
>
>It may make sense to upgrade the [go directive](https://go.dev/ref/mod#go-mod-file-go) in go.mod files to match:
>
>https://github.com/sigstore/sigstore/blob/589100556a55096e80e6b3c0ca6028da6fb9adbb/go.mod#L3

Go 1.18 was released ~a year ago; this PR brings Sigstore onto the latest stable version of Go. 

I tested this PR against `make` targets. E2E tests, linting, etc. succeed.

#### Release Note

Upgraded module go directives to Go 1.20.

#### Documentation

I linked the wrong repo below; I haven't identified any documentation changes needed.

~I think this only requires one update:~

~- https://docs.sigstore.dev/cosign/installation/#with-go-119 should read `With Go 1.20`~

~I checked by searching "1.17", "1.18", "1.19", and "1.20" in the [sigstore/docs]( https://github.com/sigstore/docs) repo.~

### CODE REVIEW

**Review by haydentherapper — CHANGES_REQUESTED** [review]
Since we aren’t using any features of go 1.20, we had decided to keep it at 1.18

**Comment by clfs** [comment]
@haydentherapper fair enough, thanks.

File: hack/tools/go.mod [hunk 1]
```diff
@@ -1,6 +1,6 @@
 module github.com/sigstore/sigstore/hack/tools
 
-go 1.17
+go 1.20
```
Comment (cpanato) [inline hack/tools/go.mod:3]: i think here we can bump to go1.20 with no problem

File: test/fuzz/go.mod [hunk 2]
```diff
@@ -1,6 +1,6 @@
 module github.com/sigstore/sigstore/test/fuzz
 
-go 1.18
+go 1.20
```
Comment (cpanato) [inline test/fuzz/go.mod:3]: here as well

File: go.mod [hunk 3]
```diff
@@ -1,6 +1,6 @@
 module github.com/sigstore/sigstore
 
-go 1.18
+go 1.20
```
Comment (cpanato) [inline go.mod:3]: echoing @haydentherapper and this means the minimum required version to build this project

I would keep this as is for now and maybe we bump to go1.19 in a near future

**Review by cpanato — CHANGES_REQUESTED** [review]

**Comment by clfs** [comment]
I reverted the top-level `go.mod` file to Go 1.18 per @cpanato's feedback.

**Review by haydentherapper — COMMENTED** [review]
I'd prefer we keep the same version of Go for the whole library, otherwise development gets tricky if you only have the older version

**Comment by cpanato** [comment]
sgtm to me as well @haydentherapper 
@clfs can you sync all in the same? thanks!

**Comment by clfs** [comment]
All `go.mod` files now sit on Go 1.18. Sorry for not checking beforehand; I could have made this smoother :)

**Review by haydentherapper — APPROVED** [review]
Sorry for not approving earlier, looks good, thanks!

**Pull request merged**